### ISSUE TITLE
Fuse softmax and masking in MultiheadAttention

### ISSUE THREAD

**Issue by alexeyr** [pytorch/pytorch#44945]
## 🚀 Feature

Fuse `softmax` and mask calculation in `multi_head_attention_forward`.

## Motivation

Currently `torch.nn.functional.multi_head_attention_forward` masks attention and then calls `softmax` as follows:

https://github.com/pytorch/pytorch/blob/caea1adc35404b39fe4f2e3fa75f9a0b9e554bbc/torch/nn/functional.py#L4265-L4281

(the code in C++ frontend is similar). In our own C++ implementation (not currently publicly available), we found that fusing these operations led to ~5% perfomance improvement in attention. Of course the number can't be expected to be the same in Pytorch, but it is probably a reasonable estimate.

## Pitch

Add a `masked_softmax` operation which effectively only calls `softmax` on un-masked values and fills the masked values with 0 without doing extra calculation. Include a fallback implementation for the cases we can't optimize (e.g. because an external library `softmax` is used, CUDA, etc.).

If accepted, we'll be happy to work on this.

The operation probably shouldn't be part of the public API, unless there are other significant uses for it.

**Comment by gchanan** [comment]
CC @zhangguanheng66

**Comment by zhangguanheng66** [comment]
Link a relevant PR for this https://github.com/pytorch/pytorch/pull/42323

**Comment by wgale** [comment]
> Link a relevant PR for this #42323

I think this would be a better way of addressing #42323 as that is trying to analyze where softmax will fail and then correct for it, whereas this tries to avoid it altogether.

> (the code in C++ frontend is similar). In our own C++ implementation (not currently publicly available), we found that fusing these operations led to ~5% perfomance improvement in attention. Of course the number can't be expected to be the same in Pytorch, but it is probably a reasonable estimate.

I've found that fusing them gave a similar performance improvement in python/pytorch (following a similar implementation to torchtext). It also makes it easier to handle #42323.

**Comment by rodosingh** [comment]
Hi everyone,
Recently, I've tried to fix this issue without going for `-inf` replacement. Very elaborately, I tried to explain and also uploaded the **modified** basic scripts, which one can use readily for their experiments.
This is the link to the [comment](https://github.com/pytorch/pytorch/issues/41508#issuecomment-1723119580) located in Thread #41508.

Thank You.
